no_std support (closes #11)

tarcieri opened the pull request:
Changes references from std:: to core:: where applicable, ensuring the crate builds in no_std environments.

tarcieri commented on src/hide.rs:
```diff
@@ -63,7 +63,7 @@ mod nightly {
-#[cfg(not(feature = "no_cc"))]
+#[cfg(all(not(feature = "no_cc"), feature = "std"))]
 mod cc {
     use std::os::raw::c_void;
```
Wow, using `std::os::raw::c_void` from `no_std` is one deep rabbit hole. Some decent background in this thread:

https://internals.rust-lang.org/t/solve-std-os-raw-c-void/3268/16

I guess this is the path forward, but work has not yet begun:

https://github.com/rust-lang/rust/issues/43467

An interim option is to use `libc::c_void` from the `libc` crate, but `no_std` users are presently stuck on nightly anyway for a number of reasons, so simply using the nightly features seems to me like the best solution for `no_std` users for now.

tarcieri replied:
On second thought, the `std` feature I added exists exclusively to support this. It'd be nice to not have a `std` feature at all.

An interim solution would be to change the `no_cc` feature to be a default `cc` feature (features are supposed to be additive anyway), and have an optional dependency on `libc` pulled in when the `cc` feature is enabled.

tarcieri commented:
Well, I redid this as a breaking change (see f012580f8be47e92002a9c5c32ba0896bfa80582 for an attempt to do it mostly backwards compatibly).

I changed the `no_cc` feature to a `cc` feature (cargo features should be additive anyway), and used that to gate pulling in the `libc`, as that seems like the best way to get at `c_void` for now. It seems this is a bit of a quagmire, and `std::os::raw` should be deprecated, but there isn't a good replacement now besides the `libc` crate:

https://internals.rust-lang.org/t/solve-std-os-raw-c-void/3268/16 

Apparently this is the path forward:

https://github.com/rust-lang/rust/issues/43467

cesarb commented:
You missed the documentation comments and the readme, which also reference `no_cc`.

While `void *` is more elegant (and `core` really should have a way to represent it), the C function `clear_on_drop_hide` is an implementation detail. It would be simple to change its signature to `unsigned char *clear_on_drop_hide(unsigned char *ptr)`, and on the Rust side use `u8` instead of `c_void` (although `char *` and `c_char` would be better than `unsigned char` and `u8`, but again we would need `c_char`). That way, you don't need the `libc` crate, and can use the C function variant even with `core` (the fallback is not as reliable, best to avoid it if at all possible).

tarcieri commented:
Okay, I changed `clear_on_drop_hide`'s signature to use `unsigned char *ptr` and changed the Cargo feature back to `no_cc` (I still think a positive `cc` feature would make more sense, but that's a topic for a separate PR if it can be avoided in this one)
